### ISSUE TITLE
Support defining methods of a function in multiple cells

### ISSUE THREAD

**Issue by karlwessel** [fonsp/Pluto.jl#177]
The other solution is much harder to implement consistently - i.e. being able to detect multiple methods with the same argument types (instead of just detecting multiple functions with the same name) and deleting only the methods from one cell, while leaving the others. 

All of this is possible within the scope of Pluto, but not something for the next couple of months :)

_Originally posted by @fonsp in https://github.com/fonsp/Pluto.jl/issues/109#issuecomment-628060038_

**Comment by karlwessel** [comment]
done, although maybe now the comments seem a bit weird :)

**Comment by fonsp** [comment]
(split into two issues: #178)

**Comment by fonsp** [comment]
It would be helpful if someone with an eye for edge cases (@karlwessel) would write the test cases for this _before_ thinking up a solution.

([Writing test cases is fairly easy,](https://github.com/fonsp/Pluto.jl/blob/master/test/React.jl#L202-L253) and they don't need to be perfect yet.)

**Comment by fonsp** [comment]
Strictly speaking, two cells `f(x::Int) = 1` & `f(x::Int) = 2` should not be allowed. _However_, there are some very difficult situations like:

`f(x::Int) = 1` & `f(x::Int64) = 2`

`A = Number` & `f(x::A) = 1` & `f(x::Number) = 2`

`A = rand(Bool) ? Number : String` & `f(x::A) = 1` & `f(x::Number) = 2`

If all of these situations are simply ignored (we probably will), then a notebook like the first situation would be _ambiguous_  (ignoring the SECRET fact that the _order in which cells are placed_ is used as a tie-breaker when two cells are mutually independent.)

**Comment by karlwessel** [comment]
Is it possible to detect those cases during execution of a notebook by comparing the `methodswith` count for each function for which there are cells defining methods for? And throw a runtime exception if one cell that should increase the count doesn't because it replaced another method?

**Comment by fonsp** [comment]
That's going in the right direction! (Although my 5th example would not get detected with that method)

**Comment by karlwessel** [comment]
Well, maybe we don't have to catch _all_ corner cases...

**Comment by fonsp** [comment]
Building on your counting methods idea:

It's very nice if we can keep all of the syntax analysis in one place, _before_ running the code, and then run cells in a precomputed order. Checking the number of defined methods after running cells (in the middle of a reactive run), and then taking that information back to do something special makes that tricky. But there are three reasons to do this kind of "dynamic" code analysis instead of static anaylsis:

- this problem
- https://github.com/fonsp/Pluto.jl/issues/24
- https://github.com/fonsp/Pluto.jl/issues/196

The common hurdles in these three is:
- the dependency graph can only be determined after runtime
- the dependency graph depends on the global state
- we'd like to know things about external packages without
  - analysing their source code, nor
  - importing them into the server process

[part 1/2]

**Comment by fonsp** [comment]
We can simplify the `do_signatures_overlap` quite a bit by dropping support for:
`A = B`
`f(x::A) = 1`
`f(x::B) = 2`

Although it is common to have user-defined types, I hope that it's uncommon for user defined _aliases_ to cause this kind of trouble. So:

## `do_signatures_overlap`?
Create a new test module, ~import all globals into the test module~, **create dummy types for all unknown types**, and then define dummy functions with the given signatures. If the number of methods is less than the number of signatures, then they overlap.

1. mini syntax analysis to find `using` (in server process)
2. call the `using` statements (in worker process)
3. syntax analysis (in server process), with `symbols_exported_by_module` and `do_signatures_overlap` to fill in the gaps
4. reactive run (in worker process)

**Comment by karlwessel** [comment]
> `f(x::Int) = 1` & `f(x::Int64) = 2` (built in type synonyms)
> 
> `A = Number` & `f(x::A) = 1` & `f(x::Number) = 2` (dynamic types)
> 
> `A = rand(Bool) ? Number : String` & `f(x::A) = 1` & `f(x::Number) = 2` (aaah)
> 
> `f = 1` & `f(x) = 1` (variables vs methods)
> 
> `f(x::Int) = 1; f(x::String) = 2` && `f(x::String) = 3; f(x::Vector) = 4` (multiple methods per cell)
> 
> `f(x::Int) = 1; f(x::Int) = 2` && `f(x::Vector) = 3` (number of methods in code might not equal number of defined methods)
> 
> `f() = 1` && `f(; s) = 2` (kwargs)
> 
> `f(x::T) where T = 1` && `f(x) = 2`
> 
> `f(x::(rand(Bool) ? Int64 : Float64)) = 1` &
> `f(x::(rand(Bool) ? Int64 : Float64)) = 2`

I guess I start implementing tests for those :)

**Comment by fonsp** [comment]
Yes please! Although I think we will be going with a solution where half of those tests are broken :)

**Comment by karlwessel** [comment]
The tests for stacktraces on the master branch are currently failing for me, is this "intentional"?

One of the failed test messages:
```
Stack traces: Test Failed at /home/karl/Pluto.jl/test/React.jl:155
  Expression: ((st["stacktrace"])[4])["line"] == 1
   Evaluated: 111 == 1
Stacktrace:
 [1] macro expansion at /home/karl/Pluto.jl/test/React.jl:155 [inlined]
 [2] macro expansion at /build/julia/src/julia-1.5.1/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115 [inlined]
 [3] macro expansion at /home/karl/Pluto.jl/test/React.jl:148 [inlined]
 [4] macro expansion at /build/julia/src/julia-1.5.1/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115 [inlined]
 [5] (::var"#111#139")() at /home/karl/Pluto.jl/test/React.jl:8
Stack traces: Test Failed at /home/karl/Pluto.jl/test/React.jl:156
```

**Comment by karlwessel** [comment]
> Yes please! Although I think we will be going with a solution where half of those tests are broken :)

Why is that? Is the method-counting version to complicated?

**Comment by fonsp** [comment]
Hm what Julia version do you use?

**Comment by karlwessel** [comment]
1.5.1

**Comment by fonsp** [comment]
> Why is that? Is the method-counting version to complicated?

Counting methods is not too complicated, but then there needs to be a feedback loop: too few methods were detected, so we error those cells manually, and then go back and find dependent cells and run those again.

Instead of:
Syntax analysis > topological ordering > execution

where each of those steps does not depend on the results of later steps

**Comment by fonsp** [comment]
Wait let me rethink that

**Comment by fonsp** [comment]
Wait let me rethink that

**Comment by Pangoraw** [comment]
Hi!
I made a little [proof of concept](https://gist.github.com/Pangoraw/0fe42c93bd8d5797654f50e9578de271) to solve this issue using a shallow module where all the types referenced are redefined using `struct var"$name" end` and where the functions have an empty body (following @fonsp's idea). Iterating on @karlwessel's idea of counting the number of methods defined, we visit the methods table to extract all defined methods for a function.  We can then reduce by using the signature as a key to identify which signature is defined more than once. This prevent one problem I have identified with counting: We couldn't have overridden method from the `Base` package for example since the number of methods would have stayed the same.

The complete process would be:
- [x] Collect the function signatures in `ExpressionExplorer` so that the symbols states contains the type signature and not just the references.
- [x] Before a reactive run if a function name is assigned in multiple cells we can construct this shallow module with a empty definition for every type.
- [x] Define a `get_errors` function which can identify which methods are duplicated
 - [x] For every function name which is assigned more than once run `get_errors` to identify whether or not there are duplicate methods
 - [x] Match the duplicate methods with their respective cells using a `LineNumberNode` containing the cell id for each shallow function
 - [x] Dispatch errors for those cells
 - [x] Remove previously dispatched `MultipleDefinitions` errors, `topology[cell].assignments` will no longer contains symbols for function names

#### What don't work yet

- [ ] Handle curly types `Union{A, B}`, `Option{A}` and so on...
- [ ] Handle optional parameters `a::Bool=false`
- [ ] Handle `::Any` Parameters (parameters without types) -> `f(a) === f(a::Any)`
- [ ] Handle type aliases `A = Int64`
- [ ] Currently I am "_polluting_" the `Main` module with these shallow modules, using another process with `Distributed` will be cleaner
- [ ] Currently, the detection is run every time there is a change on **every** cells, there must be a way to filter every function that has not changed
- [ ] Plenty of tests -> [#471](https://github.com/fonsp/Pluto.jl/pull/471)

#### A quick demo of the progress

![methods_detection](https://user-images.githubusercontent.com/9824244/95017471-c7455d80-0659-11eb-9bbc-bc56b737399b.gif)

I have pushed my changes on [this branch](https://github.com/Pangoraw/Pluto.jl/tree/duplicate-methods) if you want to take a look at the implementation and [compare the changes](https://github.com/fonsp/Pluto.jl/compare/master...Pangoraw:duplicate-methods?expand=1). If the methods implemented seems good to you I can open a draft PR to discuss more technical details of the implementation :sparkler:

**Comment by fonsp** [comment]
Hi @Pangoraw!

Thanks so much for your contribution! I've sent you an email in case you would like to call!

Your work looks promising, but motivated by @karlwessel (https://github.com/fonsp/Pluto.jl/issues/177#issuecomment-645629771) I think we should take a different direction:
These tricky corner cases are very difficult to solve, and will probably make Pluto's internals much more complicated (if we go with the crazy solution I proposed, although I see some good middle ground solutions).

So instead, I think we should only attempt to catch super simple cases like `f(x) = 1` vs `f(wow::Any) = 2`, and leave anything more complicated for later.

I'm also not sure whether I like having the function signature info in its `SymbolState`. I think it would be better to store it in its name. I have started working on this in a new branch, so far it's just refactoring old code.

During my refactoring, I found that a bunch of tricky code can be eliminated! https://github.com/fonsp/Pluto.jl/pull/536
